Upgrade postgres #9203 ke (#690)
SUSE/doc-susemanager: modules/upgrade/pages/db-migration.adoc

[[db-migration]]
= Database Migration

This is about PostgreSQL database migration.

> If you want to upgrade {productname} 3.2 to {productname} 4 you must check whether {productname} 3.2 is already based PostgreSQL version 10.

Reviewer: ```suggestion
If you want to upgrade {productname} 3.2 to {productname} 4 you must check whether {productname} 3.2 is already based on PostgreSQL version 10.
```

Reviewer 2: I'm not a native English speaker but I always thought that "must" is considered impolite and one should use "you will need to" instead. But if it has Lana's blessing it will be ok :)

Author: @mantel This is technical documentation (not marketing speech), Hubert!
We probably should define somewhere what it actually means if we say "must" or "should".

> If yes, you can skip this section.

Reviewer: ```suggestion
If you are already using PostreSQL version 10, skip this section.
```

> If it is based on a older one such as version 9.6 you must migrate PostgreSQL to version 10 first.

Reviewer: ```suggestion
If you are using an older version, such as version 9.6, you must migrate PostgreSQL to version 10 before you begin the database migration.
```

For more information about migrating PostgreSQL 9.6 to version 10, see xref:db-migration.adoc#db-migration-prepare[].

> If you still run PostgreSQL 9.4 on {productname} 3.2 see the 3.2 product documentation at https://www.suse.com/documentation/suse-manager-3/3.2/susemanager-best-practices/html/book.suma.best.practices/bp.sp.migration.html#sp.migration.postgresql.

Reviewer: ```suggestion
If you run PostgreSQL{nbsp}9.4 on {productname} 3.2, see the 3.2 product documentation: https://www.suse.com/documentation/suse-manager-3/3.2/susemanager-best-practices/html/book.suma.best.practices/bp.sp.migration.html#sp.migration.postgresql.
```

Reviewer 2: I wonder whether we should mention 9.4 at all in the first place. Quite some time ago we introduced some change that requires a "new" feature of postgresql that was not available in 9.4. So if somebody is still running this ancient postgresql version, he definitely also is running an totally outdated version of SUSE Manager. I think 9.4 is unusable for SUSE Manager since longer than a year.

Author: I see but I nevertheless would like to mention this case.  Of course, if developers or supporter would think this will cause more confusion than help, I can remove it.

Reviewer 2: Honestly I do not care. IMHO it is just a no-op by now. I bet nobody is using 9.4 any longer.

> Direct migration from PostgreSQL 9.4 to version 10 is not supported.

Reviewer: ```suggestion
You cannot migrate directly from PostgreSQL{nbsp}9.4 to version 10.
```

[[db-migration-prepare]]
== Prepare to Upgrade

Before you begin the upgrade, prepare your existing 3.2 system and create a database backup.

PostgreSQL stores data at [path]``/var/lib/pgsql/data/``, and logs to [path]``/var/lib/pgsql/data/pg_xlog/``.

.Procedure: Preparing to Upgrade

. Check the active PostgreSQL version:
+
----
psql --version
----
+
If you are using PostgreSQL{nbsp}9.6, you can upgrade to PostgreSQL{nbsp}10.

. Check the active smdba version:
+
----
rpm -q smdba
----
+
PostgreSQL{nbsp}10 requires ``smdba`` version 1.6.2 or later.

. Perform a database backup.
For more information on backing up, see xref:administration:backup-restore.adoc[].

== Upgrade PostgreSQL

[WARNING]
====
Always create a database backup before performing a migration.
For more information about a database backup, see xref:db-migration.adoc#db-migration-prepare[].
====

PostgreSQL upgrades can be performed in two ways: a regular upgrade, or a fast upgrade:

> A *regular upgrade* will create a complete copy of the database, so you will need double the existing database size of space available.

Reviewer: I'm not a fan of the emphasis here, can we write around it? Perhaps use a def list or something?

Author: agreed. i remove it.

> A *regular upgrade* will create a complete copy of the database, so you will need double the existing database size of space available.

Author: ```suggestion
A regular upgrade will create a complete copy of the database, so you will need double the existing database size of space available.
```

Regular upgrades can take a considerable amount of time, depending on the size of the database and the speed of the storage system.
Regular upgrades also require database downtime, your database will not be accessible while the upgrade is running.

A fast upgrade only takes a few minutes, and uses no additional disk space.

> However, fast upgrades are more likely to fail, and if this occurs you must restore the database from the backup.

Author: ```suggestion
However, if a fast upgrade fails, you must restore the database from the backup.
With the fast upgrade the risk to run out of disk space is smaller.

In case of failure you do not have the implicit backup that you get when doing a regular upgrade.
The regular upgrade will copy the database files instead of hardlinking them.
This statement will scare customers and they will tend to unnecessarily waste time and space by avoiding the fast upgrade.
```

> A *fast upgrade* only takes a few minutes, and uses no additional disk space.

Author: ```suggestion
A fast upgrade only takes a few minutes, and uses no additional disk space.
```

> However, fast upgrades are more likely to fail, and if this occurs you must restore the database from the backup.

Reviewer 2: No, this is utterly wrong and needs to get changed! The fast upgrade is NOT "more likely to fail", in fact it is even more safe as your risk to run out of disk space is smaller. However it is more dangerous in the sense that in case of failure you do not have the implicit backup you get when doing a regular upgrade (which will copy the database files instead of hardlinking them).
This statement will scare customers and they will tend to unnecessarily waste time and space by avoiding the fast upgrade.

Author: thanks for clarification, Hubert.  I hope it is better worded now.

. Perform a database backup.
For more information on backing up, see xref:administration:backup-restore.adoc[].

== Upgrade PostgreSQL

[WARNING]
====
Always create a database backup before performing a migration.
For more information about a database backup, see xref:db-migration.adoc#db-migration-prepare[].
====

PostgreSQL upgrades can be performed in two ways: a regular upgrade, or a fast upgrade:

A regular upgrade will create a complete copy of the database, so you will need double the existing database size of space available.
Regular upgrades can take a considerable amount of time, depending on the size of the database and the speed of the storage system.
Regular upgrades also require database downtime, your database will not be accessible while the upgrade is running.

A fast upgrade only takes a few minutes, and uses no additional disk space.
However, if a fast upgrade fails, you must restore the database from the backup.
With the fast upgrade the risk to run out of disk space is smaller.

In case of failure you do not have the implicit backup that you get when doing a regular upgrade.
The regular upgrade will copy the database files instead of hardlinking them.

> This statement will scare customers and they will tend to unnecessarily waste time and space by avoiding the fast upgrade.

Reviewer 2: This ("will scare customers") was just a comment by me. It was not meant to go into the document :)

Author: thanks for noting.  I forgot to remove it ;)

If the upgrade fails, the migration script will attempt to restore the database to its original state.
====

.Procedure: Performing a Fast PostgreSQL Upgrade
. Perform a database backup.

> Without a verified database backup, you must not initiate a fast upgrade.

Reviewer 2: Again I'm amazed by the "you must not" :) I would have worded it as "you should not" or even "it is strongly recommended..."

Author: Yes, sometimes passive voice is better.

Reviewer: We use passive voice in technical writing a lot, because the actor is often the least important information. So in this case, it's more about word order than active/passive. Here, we need the reader to notice this sentence if they don't have a backup, so we put that information first. If they have a backup, they can skip this sentence. If they don't, we need them to keep reading: do not initiate the upgrade.

As for the word choice, "should" is not as definite as "must". I often sit on the couch and say "I should clean the floors", but it doesn't mean I do. The same can be said for "it is strongly recommended". Aside from being far too wordy, it's still not as strong as "must". You don't get to argue with "must" ;)

All that said, it could be even clearer by saying "do not". This is both more concise, and more definite. But do you notice how I just removed the actor entirely? The actor is simply not important in this sentence, which further backs up the case for passive voice.

I wrote guidance on this type of thing here: https://github.com/SUSE/doc-susemanager/wiki/wug#language

Reviewer 2: Thanks for the free English lesson :)

I was under the impression that - exactly because "must" is so strong - it should not be used in English as it can be perceived as impolite and one should use something like "you might consider to do this and that" instead.

The British are too polite to be honest,
while the Germans are too honest to be polite.

:)

Reviewer: > Thanks for the free English lesson :)

Always happy to (over)share the knowledge :stuck_out_tongue: 

> 
> I was under the impression that - exactly because "must" is so strong - it should not be used in English as it can be perceived as impolite and one should use something like "you might consider to do this and that" instead.

That would probably be true if we were in a different setting. 

> 
> The British are too polite to be honest,
> while the Germans are too honest to be polite.

:joy:

And technical writers are the least polite of all. None of those extra words hanging around and messing the place up for us!
